engine/execution/storehouse/in_memory_register_store.go

```diff
@@ -0,0 +1,336 @@
+package storehouse
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/engine/execution"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+var _ execution.InMemoryRegisterStore = (*InMemoryRegisterStore)(nil)
+
+var ErrNotExecuted = fmt.Errorf("block is not executed")
+
+type ErrPruned struct {
+	PrunedHeight uint64
+	Height       uint64
+}
+
+func NewErrPruned(height uint64, prunedHeight uint64) error {
+	return ErrPruned{Height: height, PrunedHeight: prunedHeight}
+}
+
+func (e ErrPruned) Error() string {
+	return fmt.Sprintf("block is pruned at height %d", e.Height)
+}
+
+func IsErrPruned(err error) (ErrPruned, bool) {
+	var e ErrPruned
+	ok := errors.As(err, &e)
+	if ok {
+		return e, true
+	}
+	return ErrPruned{}, false
+}
+
+type InMemoryRegisterStore struct {
+	sync.RWMutex
+	registersByBlockID map[flow.Identifier]map[flow.RegisterID]flow.RegisterValue // for storing the registers
+	parentByBlockID    map[flow.Identifier]flow.Identifier                        // for register updates to be fork-aware
+	blockIDsByHeight   map[uint64]map[flow.Identifier]struct{}                    // for pruning
```

Comment:
couldn't this be of type:
```
map[uint64][]flow.Identifier
```

Reply:
When pruning conflicting forks, we need to delete a block from this index (See L313), using map allows to delete with O(1), instead of O(n) with `[]flow.Identifier`